### PR TITLE
209 switch order matching to pro rate from fifo

### PR DESCRIPTION
**Pull request by glottologist**
This PR switches out the granular FIFO (first in, first out) execution method for the more efficient pro-rata mechanism

### CODE REVIEW

File: batcher/orderbook.mligo [hunk 1]
```diff
@@ -39,87 +48,6 @@ let make_new_order (order : order) (amt: nat) : order =
...
-      let equivalent_ord_2_amount : nat = compute_equivalent_token ord_2 exchange_rate in 
-      left_partial_match equivalent_ord_2_amount ord_1 ord_2 treasury
-    | EXACT_MATCH -> 
-      total_match equivalent_ord_1_amount ord_1 ord_2 treasury
-  in 
-  match_calculation
-    
 
 [@inline]
 (*This function push orders auxording to a pro-rata "model"*)
```
Comment (Hakimba) [inline batcher/orderbook.mligo:53]: nitpicking : "auxording"

File: batcher/batcher.mligo [hunk 2]
```diff
@@ -145,17 +165,15 @@ let redeem (storage : storage) : result =
 
 
 let move_current_to_previous_if_finalized (storage : storage) : storage =
-  let batches = storage.batches in
-  let current = batches.current in
+  let batch_set = storage.batch_set in
```
Comment (kienle371999) [inline batcher/batcher.mligo:168]: I think we need to change the name of this function to `move_current_to_last_if_finalized`.

**Review by kienle371999 — CHANGES_REQUESTED** [review]

**Comment by kienle371999** [comment]
I think we need to remove the `treasury_old.mligo` from the codebase.

**Comment by glottologist** [comment]
> I think we need to remove the `treasury_old.mligo` from the codebase.

Yep, it is just there as a reference while I finish the redemption part. I will remove at the end.

**Comment by kienle371999** [comment]
I didn't see the logic of redemption for each account. Did you not add this, @glottologist?

**Comment by glottologist** [comment]
> I didn't see the logic of redemption for each account. Did you not add this, @glottologist?

Yep, that is the bit that I said I was finishing, which is why it is still WIP

**Comment by kienle371999** [comment]
@glottologist I don't see the logic of computing the tokens based on the `clearing price`. Can you tell me this?

**Comment by glottologist** [comment]
> @glottologist I don't see the logic of computing the tokens based on the `clearing price`. Can you tell me this?

get_cleared_buy_side_payout & get_cleared_sell_side_payout in the new treasury

**Comment by glottologist** [comment]
@kienle371999 @Hakimba This is ready for full review now.  Please take your time in reviewing as a lot of things have changed.

File: batcher/treasury.mligo [hunk 3]
```diff
@@ -136,178 +123,229 @@ module Utils = struct
...
+  let get_cleared_buy_side_payout
+    (swap:swap)
+    (clearing:clearing) : token_amount list =
+    let f_buy_side_actual_volume = Float.new (int clearing.prorata_equivalence.buy_side_actual_volume) 0 in
+    let f_amount = Float.new (int swap.from.amount) 0 in
+    let prorata_allocation = Float.div f_amount f_buy_side_actual_volume in
+    let f_buy_side_clearing_volume = Float.new (int (get_clearing_volume clearing)) 0 in
+    let f_sell_side_clearing_volume = Float.mul clearing.clearing_rate.rate f_buy_side_clearing_volume in
+    let payout = Float.mul prorata_allocation f_sell_side_clearing_volume in
+    let payout_equiv = Float.div clearing.clearing_rate.rate payout in
```
Comment (kienle371999) [inline batcher/treasury.mligo:209]: I think this line should be `let payout_equiv = Float.div payout clearing.clearing_rate.rate`.

**Review by kienle371999 — CHANGES_REQUESTED** [review]

File: batcher/treasury.mligo (hunk 3 above)
Comment (glottologist) [reply batcher/treasury.mligo]: Fixed

File: batcher/batcher.mligo (hunk 2 above)
Comment (glottologist) [reply batcher/batcher.mligo:168]: I am not sure that make much of a difference

**Comment by kienle371999** [comment]
LGTM.

**Pull request merged**